pkgs/build-support/trivial-builders.nix

```diff
@@ -318,23 +321,46 @@ rec {
   symlinkJoin =
     args_@{ name
          , paths
+         , outputs ? [ "out" ]
          , preferLocalBuild ? true
          , allowSubstitutes ? false
          , postBuild ? ""
          , ...
          }:
     let
+      secondaryOutputs = builtins.filter (o: o != "out") outputs;
+      outputToPathsAttr = output: {
+        name = output + "Paths";
+        value = lib.flatten (map (p: p.${output} or []) paths);
+      };
       args = removeAttrs args_ [ "name" "postBuild" ]
         // {
           inherit preferLocalBuild allowSubstitutes;
-          passAsFile = [ "paths" ];
-        }; # pass the defaults
+          outputs = [ "out" ] ++ secondaryOutputs;
+          passAsFile = [ "paths" ] ++ map (o: o + "Paths") secondaryOutputs;
+        } // builtins.listToAttrs (map outputToPathsAttr secondaryOutputs);
     in runCommand name args
       ''
         mkdir -p $out
         for i in $(cat $pathsPath); do
           ${lndir}/bin/lndir -silent $i $out
         done
+
+        for output in ${toString secondaryOutputs}; do
+          eval output_path=\$$output
+          mkdir -p $output_path
+          for i in $(eval cat \$${output}PathsPath); do
+            ${lndir}/bin/lndir -silent $i $output_path
+
+            # Replace previously propagated build outputs with our
+            # main output
+            if [ -e $output_path/nix-support/propagated-build-inputs ]; then
+              rm $output_path/nix-support/propagated-build-inputs
+              echo "$out" >> $output_path/nix-support/propagated-build-inputs
+            fi
+          done
+        done
+
```

Comment:
Is there a reason to treat `out` as special? Couldn't this logic be generalized to all outputs, whether `out` or `dev`?

Reply:
Well, there are two reasons:
1. The naming of the `paths` attribute doesn't match the naming scheme for the rest of the outputs (`*Paths`). This could be resolved by renaming it in the argument list to `runCommand`.
2. The additional outputs are treated specially in regards to propagated build outputs. If we want to generalize it, we would still need a check to make sure we don't mess with the main output's `propagated-build-inputs` file.

Reply:
> The additional outputs are treated specially in regards to propagated build outputs. If we want to generalize it, we would still need a check to make sure we don't mess with the main output's propagated-build-inputs file.

I'm unfamiliar with the intricacies of this. Should `out` inherit the `paths`' propagated build inputs? Should the secondary outputs do the same, then? Do the secondary outputs have the original `out` path as a build input? Only the original `out` path, or could they have more?

> The naming of the paths attribute doesn't match the naming scheme for the rest of the outputs (*Paths). This could be resolved by renaming it in the argument list to runCommand.

Yeah, I think the generic approach would be simpler with this, no `secondaryOutputs`

```diff
diff --git a/pkgs/build-support/trivial-builders.nix b/pkgs/build-support/trivial-builders.nix
@@ -328,7 +328,6 @@ rec {
          , ...
          }:
     let
-      secondaryOutputs = builtins.filter (o: o != "out") outputs;
       outputToPathsAttr = output: {
         name = output + "Paths";
         value = lib.flatten (map (p: p.${output} or []) paths);
@@ -336,17 +335,11 @@ rec {
       args = removeAttrs args_ [ "name" "postBuild" ]
         // {
           inherit preferLocalBuild allowSubstitutes;
-          outputs = [ "out" ] ++ secondaryOutputs;
-          passAsFile = [ "paths" ] ++ map (o: o + "Paths") secondaryOutputs;
-        } // builtins.listToAttrs (map outputToPathsAttr secondaryOutputs);
+          passAsFile = map (o: o + "Paths") outputs;
+        } // builtins.listToAttrs (map outputToPathsAttr outputs);
     in runCommand name args
       ''
-        mkdir -p $out
-        for i in $(cat $pathsPath); do
-          ${lndir}/bin/lndir -silent $i $out
-        done
-
-        for output in ${toString secondaryOutputs}; do
+        for output in ${toString outputs}; do
           eval output_path=\$$output
           mkdir -p $output_path
           for i in $(eval cat \$${output}PathsPath); do
```


Thank you for your clarifications!